### PR TITLE
[Feature:RainbowGrades] Final grade GUI configuration

### LINKED ISSUE #10443: Rainbow Grades Final Term Grade Configuration on GUI
**What problem are you trying to solve with Submitty**
Currently an instructor must manually configure and run Rainbow Grades in order to display semester grades.  This feature is not implemented in the GUI.  Here is the an example of the syntax:

```
 "display": [
        "grade_summary",
        "grade_details",
        "final_grade"
   ],
   "final_cutoff": {
        "A": 93.0,
        "A-": 90.0,
        "B+": 87.0,
        "B": 83.0,
        "B-": 80.0,
        "C+": 77.0,
        "C": 73.0,
        "C-": 70.0,
        "D+": 67.0,
        "D": 63.0
    },
```

**Describe the way you'd like to solve this problem**
Add this customization to the Rainbow Grades GUI.
Let's also add some error checking to the values.  The code may not enforce that the numbers must be equal or decrease as we go down the scale.

NOTE: We already have the "final grade" checkbox in the Display section of the GUI, but it does nothing right now. 
Screenshot 2024-05-15 at 8 51 51 PM

### PR DESCRIPTION
**Pull request by zacharymnp**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [x] Screenshots are attached to Github PR if visual/UI changes were made

Login to Instructor, then navigate to Grade Reports > Web-Based Rainbow Grades Customization. The final cutoff customization item will show/hide based on whether the final_grade checkbox under Display is toggled.
NOTE: The item will _not_ show/hide based on the final_cutoffs checkbox, which is related to displaying a histogram.

### What is the current behavior?

Resolves #10443
To display semester grades, an instructor must manually configure and run RainbowGrades. The GUI configuration is incomplete.

### What is the new behavior?
Adds GUI configuration of final grades.

image

### CODE REVIEW

File: site/app/controllers/admin/ReportController.php [hunk 1]
```diff
@@ -631,6 +631,8 @@ public function generateCustomization() {
                 'display_benchmarks' => $customization->getDisplayBenchmarks(),
                 'benchmark_percents' => (array) $customization->getBenchmarkPercent(),
                 'benchmarks_with_input_fields' => ['lowest_a-', 'lowest_b-', 'lowest_c-', 'lowest_d'],
+                'final_benchmarks_with_input_fields' => ["A", "A-", "B+", "B", "B-", "C+", "C", "C-", "D+", "D"],
```
Comment (ziesski) [inline site/app/controllers/admin/ReportController.php:634]: shouldn't this be something like `final_cutoffs_benchmark_input`?

File: site/public/js/rainbow-customization.js [hunk 2]
```diff
@@ -413,6 +413,38 @@ function getBenchmarkPercent() {
     return benchmark_percent;
 }
 
+function getFinalBenchmarkPercent() {
```
Comment (ziesski) [inline site/public/js/rainbow-customization.js:416]: 스크린샷 2024-06-17 오전 10 28 25

When I un-check final_cutoff and try to hit save changes (which means I don't want the final cutoff in my customization.json) it still would not save insisting "All final cutoffs must have a value before saving"

This needs to be addressed.

**Review by ziesski — CHANGES_REQUESTED** [review]

File: site/app/controllers/admin/ReportController.php (hunk 1 above)
Comment (zacharymnp) [reply site/app/controllers/admin/ReportController.php]: refactored names in [c51856b](https://github.com/Submitty/Submitty/pull/10518/commits/c51856bc71f31aa7c3439c4fc83bc345c3f0e652) and [3d90329](https://github.com/Submitty/Submitty/pull/10518/commits/3d903293118ff15a2f64ed912a7eff757ce40ecd)

File: site/public/js/rainbow-customization.js (hunk 2 above)
Comment (zacharymnp) [reply site/public/js/rainbow-customization.js]: should be fixed by [ab6358e](https://github.com/Submitty/Submitty/pull/10518/commits/ab6358e12bed099d941fe1a6bdd39d5df53bc606)

Comment (zacharymnp) [reply site/public/js/rainbow-customization.js]: Default cutoff values will be included in the JSON when the checkbox is unselected, but they should be unused with no effect on RainbowGrades.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 3]
```diff
@@ -73,6 +73,22 @@
...
+                <p class="rg_info_message">You may adjust the minimum percent required to receive certain final letter grades.  Enter this percentage as a decimal number out of 100.<br />As an example 90% would be entered as 90.0</p>
+                {% for cutoff_input in final_cutoff_input_fields %}
+                    {% if final_cutoff[cutoff_input] is defined %}
+                        {% set percent = final_cutoff[cutoff_input] %}
+                    {% else %}
+                        {% set percent = '' %}
+                    {% endif %}
+
+                    <p>
+                        <label for="cutoff_{{ cutoff_input }}" class="{{ cutoff_input }}">{{ cutoff_input }}</label>
```
[the remark this answers is not in the document]
Comment (zacharymnp) [reply site/app/templates/admin/RainbowCustomization.twig]: aligned in [55e88ba](https://github.com/Submitty/Submitty/pull/10518/commits/55e88bab6072db135fedfa897650791146c940f1)

**Review by ziesski — CHANGES_REQUESTED** [review]
screenshot

This is minor issue, but users would be confused so we should fix it. 
As you can see from the screenshot, when I change the cutoff, it does not detect any changes status being "No changes to save". 
Please utilize `displayChangeDetectedMessage()` function from js to address this issue accordingly.

**Comment by zacharymnp** [comment]
Saved changes will now be detected with [3eba304](https://github.com/Submitty/Submitty/pull/10518/commits/3eba3044c60bd4a91cb2620a1268f99dea682d85). A note for later: Benchmark Percents also don't detect changes; this could be fixed in another PR.

File: site/app/templates/admin/RainbowCustomization.twig (hunk 3 above)
[the remark this answers is not in the document]
Comment (zacharymnp) [reply site/app/templates/admin/RainbowCustomization.twig]: other text boxes are also unaligned, although they are less conspicuous about it.

**Review by oliiso — COMMENTED** [review]
6B4A06AA-BD10-499D-8B35-BF03C3C30C34

Something minor, but it could be nice to update the description in italics to the right of the final_grade checkbox to something like "Display final student grade and adjust percent required to receive certain letter grades" to make it more intuitive as to what its function is.

**Review by ziesski — APPROVED** [review]
Tested, works, code looks fine other then weird php-lint casting.

**Review by oliiso — APPROVED** [review]
Ready to merge, updated description of the final_grade checkbox looks good.

**Pull request merged**